src/programs/bamdownsamplerandom.1
.TH BAMDOWNSAMPLERANDOM 1 "October 2014" BIOBAMBAM
.SH NAME
bamdownsamplerandom - downsample a SAM, BAM or CRAM file
.SH SYNOPSIS
.PP
.B bamdownsamplerandom
[options]
.SH DESCRIPTION
bamdownsamplerandom reads a SAM, BAM or CRAM file from standard input,
randomly discards reads and writes the remaining reads to standard output in
BAM format. For a pair of reads either both ends are discarded or both ends
are kept. The order of reads in the output file may be different from the
order in the input if the reads in the input file are not collated by their read
name.
.PP
The following key=value pairs can be given:
.PP
.B p=<1>:
probability for a pair of reads or a single end read to be kept. By default
all reads are kept.
.PP
.B seed=<>:
seed used for the random number generator. By default the current time is
used, i.e. each run of the program will select a different subset of reads
from an input file. If the behaviour of the program needs to be
reproducible a fixed number can be used as the random seed.
.PP
.B I=<stdin>:
input file name (data is read from standard input if this option is not given)
.PP
.B inputformat=<bam>: input file format
All versions of bamtofastq come with support for the BAM input format. If
the program in addition is linked to the io_lib package, then the following
options are valid:
.IP bam:
BAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP sam:
SAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP cram:
CRAM (see http://www.ebi.ac.uk/ena/about/cram_toolkit)
.PP
.B level=<-1|0|1|9|11>:
set compression level of the output BAM file. Valid
values are
.IP -1:
zlib/gzip default compression level
.IP 0:
uncompressed
.IP 1:
zlib/gzip level 1 (fast) compression
.IP 9:
zlib/gzip level 9 (best) compression
.P
If libmaus has been compiled with support for igzip (see
https://software.intel.com/en-us/articles/igzip-a-high-performance-deflate-compressor-with-optimizations-for-genomic-data)
then an additional valid value is
.IP 11:
igzip compression
.PP
.B exclude=<SECONDARY,SUPPLEMENTARY>:
Do not include reads in the output that have any of the given flags set. The
flags are given separated by commas. Valid flags are:
.IP PAIRED:
read was paired in sequencing
.IP PROPER_PAIR:
read has been mapped as part of a proper pair
.IP UNMAP:
read was not mapped
.IP MUNMAP:
mate of read was not mapped
.IP REVERSE:
read was mapped to the reverse strand
.IP MREVERSE:
mate of read was mapped to the reverse strand
.IP READ1:
read was first read of a pair during sequencing
.IP READ2:
read was second read of a pair during sequencing
.IP SECONDARY:
alignment is secondary, i.e. an alternative mapping to the primary alignment in the same file
.IP QCFAIL:
read as marked as having failed quality control
.IP DUP:
read is marked as a duplicate of another read in the same file (see bammarkduplicates)
.IP SUPPLEMENTARY:
read is marked as supplementary alignment
.PP
.B disablevalidation=<0>:
Valid values are
.IP 0:
run input file validation on alignments (this is the default)
.IP 1:
do not check the validity of the input file (this may help for some broken
input files, but it is a security risk as it can lead to the execution of
arbitrary code through a forged input file).
.PP
.B colhlog=<18>
base two logarithm of the size of the hash table used for collation (the
default value is 18 and should work reasonably well for most input files.
Please see the biobambam paper at arxiv.org/abs/1306.0836 for details).
.PP
.B colsbs=<128M>
size of hash table overflow list in bytes (the default is 128MB and should
work reasonably well for most input files. Please see the biobambam paper at 
arxiv.org/abs/1306.0836 for details).
.PP
.B T=<bamdownsamplerandom_hostname_pid_time>
file name of temporary file used for collation
.PP
.B ranges=<>:
coordinate ranges selected from input. This option is only available for
input files in BAM format which have a corresponding index (.bai file) and
if input is via file (i.e. the I argument is set).
Valid ranges consist either of
.IP "whole\ reference\ sequence:"
a whole reference sequence (e.g. "chr1")
.IP "half\ open\ interval\ on\ reference\ sequence:"
an interval on a reference sequence half open on the right (e.g. "chr1:50000"
which means alignments overlapping chr1 from position 50000 to the end of chr1)
.IP "interval\ on\ reference\ sequence:"
an interval on a reference sequence (e.g. "chr1:50000-60000" which means
alignments overlapping positions 50000 to 60000 on chr1)
.PP
Multiple ranges are separated by space characters (e.g. ranges="chr1:10000-20000 chr1:30000-40000"). 
.PP
.B reference=: 
file name of the reference for CRAM input files. If this key is unset, then
the CRAM file header will be scanned for obtaining a reference file name.
.PP
.B tmpfile=<filename>: 
prefix for temporary files. By default the temporary files are created in the current directory
.PP
.B outputformat=<bam>: output file format.
All versions of bamsort come with support for the BAM output format. If
the program in addition is linked to the io_lib package, then the following
options are valid:
.IP bam:
BAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP sam:
SAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP cram:
CRAM (see http://www.ebi.ac.uk/ena/about/cram_toolkit). This format is not advisable for data sorted by query name.
.PP
.B O=<[stdout]>: 
output filename, standard output if unset.
.PP
.B outputthreads=<[1]>:
output helper threads, only valid for outputformat=bam.
.PP
.B md5=<0|1>:
md5 checksum creation for output file. This option can only be given if
outputformat=bam. Then valid values are
.IP 0:
do not compute checksum. This is the default.
.IP 1:
compute checksum. If the md5filename key is set, then the checksum is
written to the given file. If md5filename is unset, then no checksum will be computed.
.PP
.B md5filename
file name for md5 checksum if md5=1.
.PP
.B index=<0|1>:
compute BAM index for output file. This option can only be given if
outputformat=bam. Then valid values are
.IP 0:
do not compute BAM index. This is the default.
.IP 1:
compute BAM index. If the indexfilename key is set, then the BAM index is
written to the given file. If indexfilename is unset, then no BAM index will be computed.
.PP
.B indexfilename
file name for output BAM index if index=1.
.PP
.B hash=<0|1>:
use hash of query name instead of a random number for selection. This makes
the output depend on how random the hashes produced for the query names are,
but it has the advantage of not requiring collation to keep pairs together.
In contast the order of retained reads does not change for hash=1.
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2014 German Tischler, \(co 2011-2014 Genome Research Limited.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.37) unstable; urgency=medium

  * zlib interface changes in libmaus2 for supporting alternative zlibs at run time 

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 07 Apr 2016 13:40:07 +0200

biobambam2 (2.0.36) unstable; urgency=medium

  * add name hash based sub sampling in bamdownsamplerandom

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Apr 2016 08:12:57 +0200

biobambam2 (2.0.35) unstable; urgency=medium

  * use factory class in ranking output for bamtofastq
  * use factory class in bamtoname
  * cleanup bamdownsamplerandom source and use factory class to instantiate alignment file decoder

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 30 Mar 2016 11:12:32 +0200

biobambam2 (2.0.34) unstable; urgency=medium

  * deduce index and md5 filenames from output file name if possible

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 29 Mar 2016 10:19:51 +0200

biobambam2 (2.0.33) unstable; urgency=medium

  * update manual page for bamsort 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 16 Mar 2016 21:30:17 +0100

biobambam2 (2.0.32) unstable; urgency=medium

  * add rmdup option in bamsort 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 16 Mar 2016 21:24:31 +0100

biobambam2 (2.0.31) unstable; urgency=medium

  * fix setting sort order field in bamsort/bamsormadup in case there is no sort order field in the input file 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 07 Mar 2016 15:05:13 +0100

biobambam2 (2.0.30) unstable; urgency=medium

  * do not exit/fail bam indexing when encountering reads with mapping coordinates but flagged as unmapped as sole reads assigned to a reference sequence

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 17 Feb 2016 10:05:51 +0100

biobambam2 (2.0.29) unstable; urgency=medium

  * add murmur3 based digest option in bamseqchksum 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 11 Feb 2016 16:10:33 +0100

biobambam2 (2.0.28) unstable; urgency=medium

  * add bamrecalculatecigar program 

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 10 Feb 2016 14:14:04 +0100

biobambam2 (2.0.27) unstable; urgency=medium

  * add fastaexplode program 

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 20 Jan 2016 15:27:26 +0100

biobambam2 (2.0.26) unstable; urgency=medium

  * update to libmaus2 2.0.114
  * Set variable to identify programs to iRODS. (by Andrew Whitwham)

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 12 Jan 2016 19:53:21 +0100

biobambam2 (2.0.25) unstable; urgency=medium

  * update to libmaus2 2.0.102 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 05 Nov 2015 15:28:29 +0100

biobambam2 (2.0.24) unstable; urgency=medium

  * update to libmaus2 2.0.99 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 18:33:56 +0100

biobambam2 (2.0.23) unstable; urgency=medium

  * update to libmaus2 2.0.98 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 10:49:14 +0100

biobambam2 (2.0.22) unstable; urgency=medium

  * implicitly concatenate multiple input files as input for bamsort 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 29 Oct 2015 09:59:48 +0100

biobambam2 (2.0.21) unstable; urgency=medium

  * fix FF and RR type alignments in streaming duplicate marking
  * add MC tag (mate cigar string) when duplicate marking support is enabled in bamsort

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 26 Oct 2015 17:12:04 +0100

biobambam2 (2.0.20) unstable; urgency=medium

  * fix marking of FF and RR type duplicate records in BAM files 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 23 Oct 2015 14:55:46 +0200

biobambam2 (2.0.19) unstable; urgency=medium

  * include updates by Andrew Whitwham for improving the cigar handling of ClipAdapters


src/programs/bamdownsamplerandom.cpp
/**
    biobambam
    Copyright (C) 2009-2013 German Tischler
    Copyright (C) 2011-2013 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/

#include <biobambam2/BamBamConfig.hpp>
#include <biobambam2/Licensing.hpp>
#include <biobambam2/AttachRank.hpp>
#include <biobambam2/ResetAlignment.hpp>

#include <iomanip>

#include <config.h>

#include <libmaus2/bambam/CircularHashCollatingBamDecoder.hpp>
#include <libmaus2/bambam/BamBlockWriterBaseFactory.hpp>
#include <libmaus2/bambam/BamToFastqOutputFileSet.hpp>
#include <libmaus2/bambam/BamWriter.hpp>
#include <libmaus2/util/TempFileRemovalContainer.hpp>
#include <libmaus2/util/MemUsage.hpp>
#include <libmaus2/bambam/ProgramHeaderLineSet.hpp>
#include <libmaus2/random/Random.hpp>
#include <libmaus2/bambam/BamMultiAlignmentDecoderFactory.hpp>
#include <libmaus2/bambam/BamStreamingMarkDuplicatesSupport.hpp>
#include <libmaus2/digest/MurmurHash3_x64_128.hpp>

static int getDefaultLevel() { return Z_DEFAULT_COMPRESSION; }
static std::string getDefaultInputFormat() { return "bam"; }
static double getDefaultProb() { return 1.0; }

#include <libmaus2/lz/BgzfDeflateOutputCallbackMD5.hpp>
#include <libmaus2/bambam/BgzfDeflateOutputCallbackBamIndex.hpp>
static int getDefaultMD5() { return 0; }
static int getDefaultIndex() { return 0; }

template<typename decoder_type>
std::string getModifiedHeaderText(decoder_type const & bamdec, libmaus2::util::ArgInfo const & arginfo, bool reset = false)
{
	libmaus2::bambam::BamHeader const & header = bamdec.getHeader();
	std::string const headertext(header.text);

	// add PG line to header
	std::string upheadtext = ::libmaus2::bambam::ProgramHeaderLineSet::addProgramLine(
		headertext,
		"bamdownsamplerandom", // ID
		"bamdownsamplerandom", // PN
		arginfo.commandline, // CL
		::libmaus2::bambam::ProgramHeaderLineSet(headertext).getLastIdInChain(), // PP
		std::string(PACKAGE_VERSION) // VN
	);

	if ( reset )
	{
		std::vector<libmaus2::bambam::HeaderLine> allheaderlines = libmaus2::bambam::HeaderLine::extractLines(upheadtext);

		std::ostringstream upheadstr;
		for ( uint64_t i = 0; i < allheaderlines.size(); ++i )
			if ( allheaderlines[i].type != "SQ" )
				upheadstr << allheaderlines[i].line << std::endl;
		upheadtext = upheadstr.str();
	}

	return upheadtext;
}

template<typename writer_type>
void runSelection(libmaus2::bambam::CircularHashCollatingBamDecoder & CHCBD, uint32_t const up, writer_type Pout)
{
	// number of alignments processed
	uint64_t cnt = 0;
	// number of bytes processed
	uint64_t bcnt = 0;
	// number of alignments written
	uint64_t ocnt = 0;
	// verbosity shift
	unsigned int const verbshift = 20;
	// clock
	libmaus2::timing::RealTimeClock rtc; rtc.start();

	libmaus2::bambam::CircularHashCollatingBamDecoder::OutputBufferEntry const * ob = 0;

	while ( (ob = CHCBD.process()) )
	{
		uint64_t const precnt = cnt;
		uint32_t const rv = ::libmaus2::random::Random::rand32();

		if ( ob->fpair )
		{
			if ( rv <= up )
			{
				Pout->writeBamBlock(ob->Da,ob->blocksizea);
				Pout->writeBamBlock(ob->Db,ob->blocksizeb);
				ocnt += 2;
			}

			cnt += 2;
			bcnt += (ob->blocksizea+ob->blocksizeb);
		}
		else if ( ob->fsingle )
		{
			if ( rv <= up )
			{
				Pout->writeBamBlock(ob->Da,ob->blocksizea);
				ocnt += 1;
			}

			cnt += 1;
			bcnt += (ob->blocksizea);
		}
		else if ( ob->forphan1 )
		{
			if ( rv <= up )
			{
				Pout->writeBamBlock(ob->Da,ob->blocksizea);
				ocnt += 1;
			}

			cnt += 1;
			bcnt += (ob->blocksizea);
		}
		else if ( ob->forphan2 )
		{
			if ( rv <= up )
			{
				Pout->writeBamBlock(ob->Da,ob->blocksizea);
				ocnt += 1;
			}

			cnt += 1;
			bcnt += (ob->blocksizea);
		}

		if ( precnt >> verbshift != cnt >> verbshift )
		{
			std::cerr
				<< "[V] "
				<< (cnt >> 20)
				<< "\t"
				<< (static_cast<double>(bcnt)/(1024.0*1024.0))/rtc.getElapsedSeconds() << "MB/s"
				<< "\t" << static_cast<double>(cnt)/rtc.getElapsedSeconds()
				<< " kept " << ocnt << " (" << static_cast<double>(ocnt)/cnt << ")"
				<< std::endl;
		}
	}

	std::cerr << "[V] " << cnt << std::endl;
}

template<typename writer_type>
void runSelectionHash(libmaus2::bambam::BamAlignmentDecoder & BAD, uint32_t const up, writer_type Pout, uint32_t const seed)
{
	// number of alignments processed
	uint64_t cnt = 0;
	// number of bytes processed
	uint64_t bcnt = 0;
	// number of alignments written
	uint64_t ocnt = 0;
	// verbosity shift
	unsigned int const verbshift = 20;
	// clock
	libmaus2::timing::RealTimeClock rtc; rtc.start();

	libmaus2::bambam::BamAlignment const & algn = BAD.getAlignment();

	libmaus2::digest::MurmurHash3_x64_128 dig;
	std::size_t const dlen = dig.getDigestLength();
	libmaus2::autoarray::AutoArray<uint8_t> D(dlen);

	while ( BAD.readAlignment() )
	{
		uint64_t const precnt = cnt;
		char const * qname = algn.getName();

		dig.init(seed);
		dig.update(reinterpret_cast<uint8_t const *>(qname),strlen(qname));
		dig.digest(D.begin());

		uint32_t rv = 0;
		for ( uint64_t i = 0; i < dlen; ++i )
			rv ^= static_cast<uint32_t>(D[i]) << (8*(i&3));

		if ( rv <= up )
		{
			Pout->writeAlignment(algn);
			ocnt += 1;
		}

		cnt += 1;
		bcnt += algn.blocksize + sizeof(uint32_t);

		if ( precnt >> verbshift != cnt >> verbshift )
		{
			std::cerr
				<< "[V] "
				<< (cnt >> 20)
				<< "\t"
				<< (static_cast<double>(bcnt)/(1024.0*1024.0))/rtc.getElapsedSeconds() << "MB/s"
				<< "\t" << static_cast<double>(cnt)/rtc.getElapsedSeconds()
				<< " kept " << ocnt << " (" << static_cast<double>(ocnt)/cnt << ")"
				<< std::endl;
		}
	}

	std::cerr << "[V] " << cnt << std::endl;
}

void bamdownsamplerandom(
	libmaus2::util::ArgInfo const & arginfo,
	libmaus2::bambam::CircularHashCollatingBamDecoder & CHCBD
)
{
	if ( arginfo.getValue<unsigned int>("disablevalidation",0) )
		CHCBD.disableValidation();

	if ( arginfo.hasArg("seed") )
	{
		uint64_t const seed = arginfo.getValue<uint64_t>("seed",0);
		libmaus2::random::Random::setup(seed);
	}
	else
	{
		libmaus2::random::Random::setup();
	}

	double const p = arginfo.getValue<double>("p",getDefaultProb());

	if ( p < 0.0 || p > 1.0 )
	{
		libmaus2::exception::LibMausException se;
		se.getStream() << "Value of p must be in [0,1] but is " << p << std::endl;
		se.finish();
		throw se;
	}

	uint32_t const up =
		( p == 1 ) ?
		std::numeric_limits<uint32_t>::max() :
		static_cast<uint32_t>(std::max(0.0,
			std::min(
				std::floor(p * static_cast<double>(std::numeric_limits<uint32_t>::max()) + 0.5),
				static_cast<double>(std::numeric_limits<uint32_t>::max())
			)
		))
		;


	// construct new header
	::libmaus2::bambam::BamHeader uphead(getModifiedHeaderText(CHCBD,arginfo));
	uphead.changeSortOrder("unknown");

	/*
	 * start index/md5 callbacks
	 */
	std::string const tmpfilenamebase = arginfo.getValue<std::string>("T",arginfo.getDefaultTmpFileName());
	std::string const tmpfileindex = tmpfilenamebase + "_index";
	::libmaus2::util::TempFileRemovalContainer::addTempFile(tmpfileindex);


	std::string md5filename;
	std::string indexfilename;

	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > cbs;
	::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Pmd5cb;
	if ( arginfo.getValue<unsigned int>("md5",getDefaultMD5()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo) != std::string() )
			md5filename = libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo);
		else
			std::cerr << "[V] no filename for md5 given, not creating hash" << std::endl;

		if ( md5filename.size() )
		{
			::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Tmd5cb(new ::libmaus2::lz::BgzfDeflateOutputCallbackMD5);
			Pmd5cb = UNIQUE_PTR_MOVE(Tmd5cb);
			cbs.push_back(Pmd5cb.get());
		}
	}
	libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Pindex;
	if ( arginfo.getValue<unsigned int>("index",getDefaultIndex()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo) != std::string() )
			indexfilename = libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo);
		else
			std::cerr << "[V] no filename for index given, not creating index" << std::endl;

		if ( indexfilename.size() )
		{
			libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Tindex(new libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex(tmpfileindex));
			Pindex = UNIQUE_PTR_MOVE(Tindex);
			cbs.push_back(Pindex.get());
		}
	}
	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > * Pcbs = 0;
	if ( cbs.size() )
		Pcbs = &cbs;
	/*
	 * end md5/index callbacks
	 */

	// construct writer
	libmaus2::bambam::BamBlockWriterBase::unique_ptr_type Pout (
		libmaus2::bambam::BamBlockWriterBaseFactory::construct(uphead, arginfo, Pcbs)
	);


	runSelection(CHCBD,up,Pout.get());

	Pout.reset();

	if ( Pmd5cb )
	{
		Pmd5cb->saveDigestAsFile(md5filename);
	}
	if ( Pindex )
	{
		Pindex->flush(std::string(indexfilename));
	}
}

void bamdownsamplerandomHash(
	libmaus2::util::ArgInfo const & arginfo,
	libmaus2::bambam::BamAlignmentDecoder & BAD
)
{
	if ( arginfo.getValue<unsigned int>("disablevalidation",0) )
		BAD.disableValidation();

	if ( arginfo.hasArg("seed") )
	{
		uint64_t const seed = arginfo.getValue<uint64_t>("seed",0);
		libmaus2::random::Random::setup(seed);
	}
	else
	{
		libmaus2::random::Random::setup();
	}

	double const p = arginfo.getValue<double>("p",getDefaultProb());

	if ( p < 0.0 || p > 1.0 )
	{
		libmaus2::exception::LibMausException se;
		se.getStream() << "Value of p must be in [0,1] but is " << p << std::endl;
		se.finish();
		throw se;
	}

	uint32_t const up =
		( p == 1 ) ?
		std::numeric_limits<uint32_t>::max() :
		static_cast<uint32_t>(std::max(0.0,
			std::min(
				std::floor(p * static_cast<double>(std::numeric_limits<uint32_t>::max()) + 0.5),
				static_cast<double>(std::numeric_limits<uint32_t>::max())
			)
		))
		;


	// construct new header
	::libmaus2::bambam::BamHeader uphead(getModifiedHeaderText(BAD,arginfo));

	/*
	 * start index/md5 callbacks
	 */
	std::string const tmpfilenamebase = arginfo.getValue<std::string>("T",arginfo.getDefaultTmpFileName());
	std::string const tmpfileindex = tmpfilenamebase + "_index";
	::libmaus2::util::TempFileRemovalContainer::addTempFile(tmpfileindex);


	std::string md5filename;
	std::string indexfilename;

	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > cbs;
	::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Pmd5cb;
	if ( arginfo.getValue<unsigned int>("md5",getDefaultMD5()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo) != std::string() )
			md5filename = libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo);
		else
			std::cerr << "[V] no filename for md5 given, not creating hash" << std::endl;

		if ( md5filename.size() )
		{
			::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Tmd5cb(new ::libmaus2::lz::BgzfDeflateOutputCallbackMD5);
			Pmd5cb = UNIQUE_PTR_MOVE(Tmd5cb);
			cbs.push_back(Pmd5cb.get());
		}
	}
	libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Pindex;
	if ( arginfo.getValue<unsigned int>("index",getDefaultIndex()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo) != std::string() )
			indexfilename = libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo);
		else
			std::cerr << "[V] no filename for index given, not creating index" << std::endl;

		if ( indexfilename.size() )
		{
			libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Tindex(new libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex(tmpfileindex));
			Pindex = UNIQUE_PTR_MOVE(Tindex);
			cbs.push_back(Pindex.get());
		}
	}
	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > * Pcbs = 0;
	if ( cbs.size() )
		Pcbs = &cbs;
	/*
	 * end md5/index callbacks
	 */

	// construct writer
	libmaus2::bambam::BamBlockWriterBase::unique_ptr_type Pout (
		libmaus2::bambam::BamBlockWriterBaseFactory::construct(uphead, arginfo, Pcbs)
	);


	runSelectionHash(BAD,up,Pout.get(),libmaus2::random::Random::rand32());

	Pout.reset();

	if ( Pmd5cb )
	{
		Pmd5cb->saveDigestAsFile(md5filename);
	}
	if ( Pindex )
	{
		Pindex->flush(std::string(indexfilename));
	}
}

void bamdownsamplerandom(libmaus2::util::ArgInfo const & arginfo)
{
	uint32_t const excludeflags = libmaus2::bambam::BamFlagBase::stringToFlags(arginfo.getUnparsedValue("exclude","SECONDARY,SUPPLEMENTARY"));

	libmaus2::util::TempFileRemovalContainer::setup();
	std::string const tmpfilename = arginfo.getValue<std::string>("T",arginfo.getDefaultTmpFileName());
	libmaus2::util::TempFileRemovalContainer::addTempFile(tmpfilename);

	unsigned int const hlog = arginfo.getValue<unsigned int>("colhlog",18);
	uint64_t const sbs = arginfo.getValueUnsignedNumeric<uint64_t>("colsbs",128ull*1024ull*1024ull);
	bool hash = arginfo.getValue<unsigned int>("hash",false);

	libmaus2::bambam::BamAlignmentDecoderWrapper::unique_ptr_type
		Pdecoder(libmaus2::bambam::BamMultiAlignmentDecoderFactory::construct(arginfo,false, /* put rank */NULL /* copy stream */,std::cin,false,false));

	if ( hash )
	{
		bamdownsamplerandomHash(arginfo,Pdecoder->getDecoder());
	}
	else
	{
		libmaus2::bambam::CircularHashCollatingBamDecoder CHCBD(Pdecoder->getDecoder(),tmpfilename,excludeflags,hlog,sbs);
		bamdownsamplerandom(arginfo,CHCBD);
	}

	std::cout.flush();
}

int main(int argc, char * argv[])
{
	try
	{
		libmaus2::timing::RealTimeClock rtc; rtc.start();

		::libmaus2::util::ArgInfo arginfo(argc,argv);

		if ( arginfo.hasArg("filename") )
		{
			arginfo.replaceKey("I",arginfo.getUnparsedValue("filename",std::string()));
			arginfo.removeKey("filename");
		}

		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if (
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if (
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license() << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;

				std::vector< std::pair<std::string,std::string> > V;

				V.push_back ( std::pair<std::string,std::string> ( "level=<["+::biobambam2::Licensing::formatNumber(getDefaultLevel())+"]>", libmaus2::bambam::BamBlockWriterBaseFactory::getBamOutputLevelHelpText() ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("p=<[")+libmaus2::util::NumberSerialisation::formatNumber(getDefaultProb(),0)+"]>", "probability for keeping read" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("seed=<[]>"), "random seed" ) );
				V.push_back ( std::pair<std::string,std::string> ( "I=<[stdin]>", "input filename (default: read file from standard input)" ) );
				#if defined(BIOBAMBAM_LIBMAUS2_HAVE_IO_LIB)
				V.push_back ( std::pair<std::string,std::string> ( std::string("inputformat=<[")+getDefaultInputFormat()+"]>", "input format: cram, bam or sam" ) );
				V.push_back ( std::pair<std::string,std::string> ( "reference=<[]>", "name of reference FastA in case of inputformat=cram" ) );
				#else
				V.push_back ( std::pair<std::string,std::string> ( "inputformat=<[bam]>", "input format: bam" ) );
				#endif
				V.push_back ( std::pair<std::string,std::string> ( "ranges=<[]>", "input ranges (bam input only, default: read complete file)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "exclude=<[SECONDARY,SUPPLEMENTARY]>", "exclude alignments matching any of the given flags" ) );
				V.push_back ( std::pair<std::string,std::string> ( "disablevalidation=<[0]>", "disable validation of input data" ) );
				V.push_back ( std::pair<std::string,std::string> ( "colhlog=<[18]>", "base 2 logarithm of hash table size used for collation" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("colsbs=<[")+libmaus2::util::NumberSerialisation::formatNumber(128ull*1024*1024,0)+"]>", "size of hash table overflow list in bytes" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("T=<[") + arginfo.getDefaultTmpFileName() + "]>" , "temporary file name" ) );
				V.push_back ( std::pair<std::string,std::string> ( "md5=<["+::biobambam2::Licensing::formatNumber(getDefaultMD5())+"]>", "create md5 check sum (default: 0)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "md5filename=<filename>", "file name for md5 check sum (default: extend output file name)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "index=<["+::biobambam2::Licensing::formatNumber(getDefaultIndex())+"]>", "create BAM index (default: 0)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "indexfilename=<filename>", "file name for BAM index file (default: extend output file name)" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("outputformat=<[")+libmaus2::bambam::BamBlockWriterBaseFactory::getDefaultOutputFormat()+"]>", std::string("output format (") + libmaus2::bambam::BamBlockWriterBaseFactory::getValidOutputFormats() + ")" ) );
				V.push_back ( std::pair<std::string,std::string> ( "outputthreads=<[1]>", "output helper threads (for outputformat=bam only, default: 1)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "O=<[stdout]>", "output filename (standard output if unset)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "hash=<[0]>", "use query name hash instead of random number for selection (default: 0)" ) );

				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;
				std::cerr << "Alignment flags: PAIRED,PROPER_PAIR,UNMAP,MUNMAP,REVERSE,MREVERSE,READ1,READ2,SECONDARY,QCFAIL,DUP,SUPPLEMENTARY" << std::endl;

				std::cerr << std::endl;
				return EXIT_SUCCESS;
			}


		bamdownsamplerandom(arginfo);

		std::cerr << "[V] " << libmaus2::util::MemUsage() << " wall clock time " << rtc.formatTime(rtc.getElapsedSeconds()) << std::endl;
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}
